internal/testrunner/runners/system/runner.go

```diff
@@ -143,40 +144,70 @@ func (r *runner) run() (results []testrunner.TestResult, err error) {
+	if !found {
+		return result.WithError(errors.New("data stream root not found"))
+	}
 
-		var partial []testrunner.TestResult
-		if testConfig.Skip == nil {
-			logger.Debugf("running test with configuration '%s'", testConfig.Name())
-			partial, err = r.runTest(testConfig, ctxt)
-		} else {
-			logger.Warnf("skipping %s test for %s/%s: %s (details: %s)",
-				TestType, r.options.TestFolder.Package, r.options.TestFolder.DataStream,
-				testConfig.Skip.Reason, testConfig.Skip.Link.String())
-			result := r.newResult(testConfig.Name())
-			partial, err = result.WithSkip(testConfig.Skip)
-		}
+	cfgFiles, err := listConfigFiles(r.options.TestFolder.Path)
+	if err != nil {
+		return result.WithError(errors.Wrap(err, "failed listing test case config cfgFiles"))
+	}
 
-		results = append(results, partial...)
-		tdErr := r.tearDownTest()
-		if err != nil {
-			return results, err
-		}
-		if tdErr != nil {
-			return results, errors.Wrap(tdErr, "failed to tear down runner")
+	devDeployPath, err := servicedeployer.FindDevDeployPath(servicedeployer.FactoryOptions{
+		PackageRootPath:    r.options.PackageRootPath,
+		DataStreamRootPath: dataStreamPath,
+	})
+	if err != nil {
+		return result.WithError(errors.Wrap(err, "_dev/deploy directory not found"))
+	}
+
+	variantsFile, err := servicedeployer.ReadVariantsFile(devDeployPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return result.WithError(errors.Wrap(err, "can't read service variant"))
+	}
+
+	for _, cfgFile := range cfgFiles {
+		for _, variantName := range anyVariants(variantsFile) {
+			serviceOptions := servicedeployer.FactoryOptions{
+				PackageRootPath:    r.options.PackageRootPath,
+				DataStreamRootPath: dataStreamPath,
+				Variant:            variantName,
+			}
+
+			var ctxt servicedeployer.ServiceContext
+			ctxt.Name = r.options.TestFolder.Package
+			ctxt.Logs.Folder.Local = locationManager.ServiceLogDir()
+			ctxt.Logs.Folder.Agent = ServiceLogsAgentDir
+			ctxt.Test.RunID = createTestRunID()
+			testConfig, err := newConfig(filepath.Join(r.options.TestFolder.Path, cfgFile), ctxt, variantName)
+			if err != nil {
+				return result.WithError(errors.Wrapf(err, "unable to load system test case file '%s'", cfgFile))
+			}
+
+			var partial []testrunner.TestResult
+			if testConfig.Skip == nil {
+				logger.Debugf("running test with configuration '%s'", testConfig.Name())
+				partial, err = r.runTest(testConfig, ctxt, serviceOptions)
+			} else {
+				logger.Warnf("skipping %s test for %s/%s: %s (details: %s)",
+					TestType, r.options.TestFolder.Package, r.options.TestFolder.DataStream,
+					testConfig.Skip.Reason, testConfig.Skip.Link.String())
+				result := r.newResult(testConfig.Name())
+				partial, err = result.WithSkip(testConfig.Skip)
+			}
+
+			results = append(results, partial...)
+			tdErr := r.tearDownTest()
+			if err != nil {
+				return results, err
+			}
+			if tdErr != nil {
+				return results, errors.Wrap(tdErr, "failed to tear down runner")
+			}
```

Comment:
Nit. Move the code in this loop to run the test on a variant to its own method?

Reply:
Good idea, fixed.